stm32-modules/thermocycler-refresh/firmware/motor_task/motor_hardware.c

```diff
@@ -0,0 +1,275 @@
+
+#define SOLENOID_Port GPIOD
+#define SOLENOID_Pin GPIO_PIN_8
+#define LID_STEPPER_ENABLE_Port GPIOE
+#define LID_STEPPER_RESET_Pin GPIO_PIN_12
+#define LID_STEPPER_ENABLE_Pin GPIO_PIN_10
+#define LID_STEPPER_FAULT_Pin GPIO_PIN_11
+#define LID_STEPPER_CONTROL_Port GPIOA
+#define LID_STEPPER_VREF_Pin GPIO_PIN_4
+#define LID_STEPPER_DIR_Pin GPIO_PIN_1
+#define LID_STEPPER_STEP_Pin GPIO_PIN_0
+#define LID_STEPPER_VREF_CHANNEL DAC_CHANNEL_1
+#define LID_STEPPER_STEP_Channel TIM_CHANNEL_1
+
+// ----------------------------------------------------------------------------
+// Local typedefs
+
+typedef struct {
+    int32_t step_count;
+    int32_t step_target;
+} motor_hardware_status;
+
+typedef struct {
+    // Handle for the lid current control DAC
+    DAC_HandleTypeDef lid_dac;
+    // Current status of the lid stepper
+    motor_hardware_status lid_stepper;
+    // Handle for the timer used for the lid stepper motor
+    TIM_HandleTypeDef lid_timer;
+    // Callback for completion of a lid stepper movement
+    void (*lid_stepper_complete)();
+} motor_hardware_handles;
+
+// Local variables
+
+static motor_hardware_handles _motor_hardware = {
+    .lid_dac = {0},
+    .lid_stepper = {0},
+    .lid_timer = {0},
+    .lid_stepper_complete = NULL
+};
+
+// ----------------------------------------------------------------------------
+// Local function declaration
+
+static void MX_GPIO_Init(void);
+void HAL_TIM_OC_MspInit(TIM_HandleTypeDef* htim);
+static void MX_DAC_Init(DAC_HandleTypeDef* hdac);
+static void MX_OC_Init(TIM_HandleTypeDef* htim);
+
+// ----------------------------------------------------------------------------
+// Public function implementation
+
+void motor_hardware_setup(const motor_hardware_callbacks* callbacks) {
+    configASSERT(callbacks != NULL);
+
+    _motor_hardware.lid_stepper_complete = callbacks->lid_stepper_complete;
+
+    MX_GPIO_Init();
+    MX_DAC_Init(&_motor_hardware.lid_dac);
+    MX_OC_Init(&_motor_hardware.lid_timer);
+    
+    motor_hardware_lid_stepper_reset();
+}
+
+//PA0/PA1/PB10/PB11 = TIM2CH1/2/3/4 (all GPIO_AF1_TIM2)
+//control via increment_step in motor_task.hpp?
+//handle end switch IT start and stop
+void motor_hardware_lid_stepper_start(int32_t steps) {
+    //check for fault
+    _motor_hardware.lid_stepper.step_count = 0;
+    // Multiply number of steps by 2 because the timer is in toggle mode (2 interrupts = 1 microstep)
+    _motor_hardware.lid_stepper.step_target = abs(steps) * 2;
+
+    if (steps > 0) {
+        HAL_GPIO_WritePin(LID_STEPPER_CONTROL_Port, LID_STEPPER_DIR_Pin, GPIO_PIN_SET);
+    } else {
+        HAL_GPIO_WritePin(LID_STEPPER_CONTROL_Port, LID_STEPPER_DIR_Pin, GPIO_PIN_RESET);
+    }
+
+    HAL_TIM_OC_Start_IT(&_motor_hardware.lid_timer, LID_STEPPER_STEP_Channel);
+}
+
+void motor_hardware_lid_stepper_stop() {
+    HAL_TIM_OC_Stop_IT(&_motor_hardware.lid_timer, LID_STEPPER_STEP_Channel);
+}
+
+void HAL_TIM_OC_DelayElapsedCallback(TIM_HandleTypeDef *htim)
```

Comment:
you know, if you're feeling your oats you could do a timer cascade here. Have a second counter whose input trigger is TIM2 channel reload, with a compare value of the number of steps, and have _that_ timer's interrupt value stop the first. Or maybe you can make that part automatic too.